### PR TITLE
SPUI-347 > Add price formatting

### PR DESCRIPTION
**Pull request by VovaMedvediev**

### CODE REVIEW

File: lib/app/shared/helpers/format_asset_price_value.dart [hunk 1]
```diff
@@ -0,0 +1,29 @@
+const _dollarPrefixSymbol = '\$';
+const _euroPrefixSymbol = '€';
+
+String formatPriceValue({
```
Comment (eli1stark) [inline lib/app/shared/helpers/format_asset_price_value.dart:4]: have questions (in grooming)

File: lib/app/shared/models/currency_model.dart [hunk 2]
```diff
@@ -8,6 +8,8 @@ part 'currency_model.freezed.dart';
 class CurrencyModel with _$CurrencyModel {
   const factory CurrencyModel({
     required String symbol,
+    required String prefixSymbol,
+    required String baseCurrencySymbol,
```
Comment (eli1stark) [inline lib/app/shared/models/currency_model.dart:12]: We have BaseCurrencyModel, mixing different structures

File: lib/app/shared/models/currency_model.dart [hunk 3]
```diff
@@ -27,6 +29,7 @@ class CurrencyModel with _$CurrencyModel {
     required double currentPrice,
     required double dayPriceChange,
     required double dayPercentChange,
+    required int baseCurrencyAccuracy,
```
Comment (eli1stark) [inline lib/app/shared/models/currency_model.dart:32]: We have BaseCurrencyModel, mixing different structures

File: lib/app/screens/market/model/market_item_model.dart [hunk 4]
```diff
@@ -17,6 +17,10 @@ class MarketItemModel with _$MarketItemModel {
     required double dayPercentChange,
     required double assetBalance,
     required double baseBalance,
+    required String prefixSymbol,
+    required String baseCurrencySymbol,
```
Comment (eli1stark) [inline lib/app/screens/market/model/market_item_model.dart:21]: We have BaseCurrencyModel, mixing different structures

File: lib/app/screens/market/model/market_item_model.dart [hunk 5]
```diff
@@ -17,6 +17,10 @@ class MarketItemModel with _$MarketItemModel {
...
+    required int accuracy,
+    required int baseCurrencyAccuracy,
```
Comment (eli1stark) [inline lib/app/screens/market/model/market_item_model.dart:23]: We have BaseCurrencyModel, mixing different structures

**Review by eli1stark — CHANGES_REQUESTED** [review]
Needs the second review after changes

File: lib/app/shared/helpers/format_asset_price_value.dart (hunk 1 above)
Comment (VovaMedvediev) [reply lib/app/shared/helpers/format_asset_price_value.dart:1]: refactored

File: lib/app/shared/models/currency_model.dart (hunk 2 above)
Comment (VovaMedvediev) [reply lib/app/shared/models/currency_model.dart]: refactored

File: lib/app/shared/models/currency_model.dart (hunk 3 above)
Comment (VovaMedvediev) [reply lib/app/shared/models/currency_model.dart]: refactored

File: lib/app/screens/market/model/market_item_model.dart (hunk 4 above)
Comment (VovaMedvediev) [reply lib/app/screens/market/model/market_item_model.dart]: refactored

File: lib/app/screens/market/model/market_item_model.dart (hunk 5 above)
Comment (VovaMedvediev) [reply lib/app/screens/market/model/market_item_model.dart]: refactored

File: lib/app/shared/helpers/format_asset_price_value.dart [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  String? prefix,
+  num? value,
+  String? symbol,
+  int? accuracy,
```
Comment (eli1stark) [inline lib/app/shared/helpers/format_asset_price_value.dart:5]: all parameters should be required, and only prefix can be null

File: lib/app/shared/helpers/format_asset_price_value.dart [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+}) {
+  if (value != null) {
+    final validatedAccuracy = accuracy ?? 2;
```
Comment (eli1stark) [inline lib/app/shared/helpers/format_asset_price_value.dart:8]: You don't need this if params are required

File: lib/app/shared/helpers/format_asset_price_value.dart [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    final absValue = value.abs().toStringAsFixed(validatedAccuracy);
+    var formattedSymbol = '';
```
Comment (eli1stark) [inline lib/app/shared/helpers/format_asset_price_value.dart:10]: You don't need this if params are required

File: lib/app/shared/helpers/format_asset_price_value.dart [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+
+    if (symbol != null) {
+      formattedSymbol = ' $symbol';
+    }
+
+    if (value.isNegative) {
+      return '-$prefix$absValue$symbol';
+    }
+    return '${prefix ?? ''}$value$formattedSymbol';
```
Comment (eli1stark) [inline lib/app/shared/helpers/format_asset_price_value.dart:19]: inconsistency in the function.
why in the negative is absValue in the positive just value? why the symbol is different?
they should behave the same

File: lib/app/shared/helpers/format_asset_price_value.dart [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+  }
+
+  return '';
```
Comment (eli1stark) [inline lib/app/shared/helpers/format_asset_price_value.dart:22]: the function can't return an empty string

**Review by eli1stark — CHANGES_REQUESTED** [review]
The ideal function should look like this:
```dart
String formatPriceValue1({
  String? prefix,
  required num value,
  required String symbol,
  required int accuracy,
}) {
  final absValue = value.abs().toStringAsFixed(accuracy);

  if (value.isNegative) {
    if (prefix == null) {
      return '-$absValue$symbol';
    } else {
      return '-$prefix$absValue';
    }
  } else {
    if (prefix == null) {
      return '$absValue$symbol';
    } else {
      return '$prefix$absValue';
    }
  }
}
```
This function should handle cases where `0.00` must look like this `0`. Right now this function doesn't handle this

File: lib/app/shared/providers/currencies_pod/currencies_pod.dart [hunk 11]
```diff
@@ -31,6 +31,7 @@ final currenciesPod = Provider.autoDispose<List<CurrencyModel>>((ref) {
           fees: asset.fees,
           withdrawalMethods: asset.withdrawalMethods,
           iconUrl: validIconUrl(asset.iconUrl),
+          prefixSymbol: asset.prefixSymbol ?? '',
```
Comment (eli1stark) [inline lib/app/shared/providers/currencies_pod/currencies_pod.dart:34]: `prefixSymbol` must be null, if it's an empty string it breaks your function

File: lib/app/screens/market/model/market_item_model.dart [hunk 12]
```diff
@@ -17,6 +17,8 @@ class MarketItemModel with _$MarketItemModel {
     required double dayPercentChange,
     required double assetBalance,
     required double baseBalance,
+    required String prefixSymbol,
```
Comment (eli1stark) [inline lib/app/screens/market/model/market_item_model.dart:20]: `prefixSymbol` must be null because it breaks your function

**Review by eli1stark — CHANGES_REQUESTED** [review]

**Comment by eli1stark** [comment]
Screenshot_1632476613
Instead of €0

File: lib/app/shared/helpers/format_asset_price_value.dart (hunk 6 above)
Comment (VovaMedvediev) [reply lib/app/shared/helpers/format_asset_price_value.dart]: fixed

File: lib/app/shared/helpers/format_asset_price_value.dart (hunk 7 above)
Comment (VovaMedvediev) [reply lib/app/shared/helpers/format_asset_price_value.dart]: +

File: lib/app/shared/helpers/format_asset_price_value.dart (hunk 8 above)
Comment (VovaMedvediev) [reply lib/app/shared/helpers/format_asset_price_value.dart]: +

File: lib/app/shared/helpers/format_asset_price_value.dart (hunk 9 above)
Comment (VovaMedvediev) [reply lib/app/shared/helpers/format_asset_price_value.dart]: fixed

File: lib/app/shared/helpers/format_asset_price_value.dart (hunk 10 above)
Comment (VovaMedvediev) [reply lib/app/shared/helpers/format_asset_price_value.dart]: fixed

File: lib/app/shared/providers/currencies_pod/currencies_pod.dart (hunk 11 above)
Comment (VovaMedvediev) [reply lib/app/shared/providers/currencies_pod/currencies_pod.dart]: fixed

File: lib/app/screens/market/model/market_item_model.dart (hunk 12 above)
Comment (VovaMedvediev) [reply lib/app/screens/market/model/market_item_model.dart]: fixed

**Review by eli1stark — APPROVED** [review]

**Pull request merged**